AudioDeviceExample/AudioDevices/ExampleCoreAudioDevice.m

```diff
@@ -0,0 +1,487 @@
+        self.renderingContext->deviceContext = context;
+        self.renderingContext->maxFramesPerBuffer = _renderingFormat.framesPerBuffer;
+
+        const NSTimeInterval sessionBufferDuration = [AVAudioSession sharedInstance].IOBufferDuration;
+        const double sessionSampleRate = [AVAudioSession sharedInstance].sampleRate;
+        const size_t sessionFramesPerBuffer = (size_t)(sessionSampleRate * sessionBufferDuration + .5);
+        self.renderingContext->expectedFramesPerBuffer = sessionFramesPerBuffer;
+
+        NSAssert(self.audioUnit == NULL, @"The audio unit should not be created yet.");
+        if (![self setupAudioUnit:self.renderingContext]) {
+            free(self.renderingContext);
+            self.renderingContext = NULL;
+            return NO;
+        }
+    }
+    return [self startAudioUnit];
+}
+
+- (BOOL)stopRendering {
+    [self stopAudioUnit];
+
+    @synchronized(self) {
+        [self teardownAudioUnit];
+
+        NSAssert(self.renderingContext != NULL, @"Should have a rendering context.");
+        free(self.renderingContext);
+        self.renderingContext = NULL;
+    }
+
+    return YES;
+}
+
+#pragma mark - TVIAudioDeviceCapturer
+
+- (nullable TVIAudioFormat *)captureFormat {
+    /*
+     * We don't support capturing and return a nil format to indicate this. The other TVIAudioDeviceCapturer methods
+     * are simply stubs.
+     */
+    return nil;
+}
+
+- (BOOL)initializeCapturer {
+    return NO;
+}
+
+- (BOOL)startCapturing:(nonnull TVIAudioDeviceContext)context {
+    return NO;
+}
+
+- (BOOL)stopCapturing {
+    return NO;
+}
+
+#pragma mark - Private (AudioUnit callbacks)
+
+static OSStatus ExampleCoreAudioDevicePlayoutCallback(void *refCon,
+                                                      AudioUnitRenderActionFlags *actionFlags,
+                                                      const AudioTimeStamp *timestamp,
+                                                      UInt32 busNumber,
+                                                      UInt32 numFrames,
+                                                      AudioBufferList *bufferList) {
+    assert(bufferList->mNumberBuffers == 1);
+    assert(bufferList->mBuffers[0].mNumberChannels <= 2);
+    assert(bufferList->mBuffers[0].mNumberChannels > 0);
+
+    ExampleCoreAudioContext *context = (ExampleCoreAudioContext *)refCon;
+    int8_t *audioBuffer = (int8_t *)bufferList->mBuffers[0].mData;
+    UInt32 audioBufferSizeInBytes = bufferList->mBuffers[0].mDataByteSize;
+
+    // Render silence if there are temporary mismatches between CoreAudio and our rendering format.
+    if (numFrames > context->maxFramesPerBuffer) {
+        NSLog(@"Can handle a max of %u frames but got %u.", (unsigned int)context->maxFramesPerBuffer, (unsigned int)numFrames);
+        *actionFlags |= kAudioUnitRenderAction_OutputIsSilence;
+        memset(audioBuffer, 0, audioBufferSizeInBytes);
+        return noErr;
+    }
```

Comment:
Should I keep the NSLog statement? This probably shouldn't be done on a real-time thread but its helpful to know if we are getting unexpected callbacks in example code.